sklearn/preprocessing/_polynomial.py

```diff
@@ -394,17 +401,55 @@ def transform(self, X):
             if self._min_degree <= 1 and self._max_degree > 0:
                 to_stack.append(X)
             for deg in range(max(2, self._min_degree), self._max_degree + 1):
-                Xp_next = _csr_polynomial_expansion(
-                    X.data, X.indices, X.indptr, X.shape[1], self.interaction_only, deg
+                # Count how many nonzero elements the expanded matrix will contain.
+                total_nnz = sum(
+                    _calc_expanded_dimensionality(
+                        X.indptr[row_i + 1] - X.indptr[row_i],
+                        self.interaction_only,
+                        deg,
+                    )
+                    for row_i in range(X.indptr.shape[0] - 1)
                 )
-                if Xp_next is None:
+                expanded_d = _calc_expanded_dimensionality(
+                    X.shape[1], self.interaction_only, deg
+                )
+
+                if expanded_d == 0:
                     break
-                to_stack.append(Xp_next)
+                assert expanded_d > 0
+                max_indices = expanded_d - 1
+                max_indptr = total_nnz - 1
+                max_int32 = np.iinfo(np.int32).max
+                if max_indices > max_int32 or max_indptr > max_int32:
+                    index_t = np.int64
+                else:
+                    index_t = np.int32
+                expanded_data = np.ndarray(shape=total_nnz, dtype=X.data.dtype)
+                expanded_indices = np.ndarray(shape=total_nnz, dtype=index_t)
+                expanded_indptr = np.ndarray(shape=X.indptr.shape[0], dtype=index_t)
+                _csr_polynomial_expansion(
+                    X.data,
+                    X.indices,
+                    X.indptr,
+                    X.shape[1],
+                    expanded_data,
+                    expanded_indices,
+                    expanded_indptr,
+                    self.interaction_only,
+                    deg,
+                )
+                to_stack.append(
+                    sparse.csr_matrix(
+                        (expanded_data, expanded_indices, expanded_indptr),
+                        shape=(X.indptr.shape[0] - 1, expanded_d),
+                        dtype=X.dtype,
+                    )
+                )
             if len(to_stack) == 0:
                 # edge case: deal with empty matrix
                 XP = sparse.csr_matrix((n_samples, 0), dtype=X.dtype)
             else:
-                XP = sparse.hstack(to_stack, format="csr")
+                XP = sparse.hstack(to_stack, format="csr", dtype=X.dtype)
```

Comment:
While working on the tests suggested by @ogrisel , I came upon an unfortunate problem:

This line is problematic, since it turns out that during the `hstack` call, all matrices in `to_stack` get first converted to `COO` matrices, which automatically creates a `col` `ndarray` of `dtype=np.int32` if small enough. 

For example, if `n_features=65535, interaction_only=False, include_bias=False, degree=2` then there are two matrices w/ `65535` and `2147450880` columns respectively. Both are smaller than `np.iinfo(np.int32).max` so the resulting `COO` matrices have `col` arrays of type `np.int32`.

The problem is then when the offset is calculated for concatenation, e.g. `2147450880+65535` is calculated, which results in an integer overflow for `np.int32` leading to negative column indices.

This looks like a solid upstream issue, and I'm not sure what the best way around this would be. Perhaps implementing our own narrow-scope `hstack`?

Reply:
Maybe we can skip those new tests when `sp_version < parse_version("1.8")` then.